### PR TITLE
V22F-708 Correct UID on startup

### PR DESCRIPTION
**Pull request by VoidLeech**
User id is now left empty on startup
Generate button has been added. This will generate a random four-digit user id. Id's are left-padded with 0's if below 1000.
Apply button will now also give an error if there is no user id, even if there was one before a user manually cleared the entire input field. Choosing a black box will not remove the user id error, nor will choosing a user id remove the error that a black box hasn't been chosen.

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 1]
```diff
@@ -535,13 +553,37 @@ static void SelectBlackBox(void* /* dummy */)
...
 
+/// @brief Generates a random user id
+static void GenerateUid(void* /* dummy */)
+{
+    // Create random userId
+    std::random_device rd;
+    std::default_random_engine generator(rd());
+    std::uniform_int_distribution<int> distribution(0, 9999);
+    // Assign it
+    sprintf(m_userId, "%04d", distribution(generator));
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:575]: Use the Random generator from rppUtils/Random.hpp

**Review by larsdekwant — CHANGES_REQUESTED** [review]

**Review by suppergerrie2 — APPROVED** [review]

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 2]
```diff
@@ -535,13 +553,36 @@ static void SelectBlackBox(void* /* dummy */)
...
     m_blackBoxChosen = true;
 }
 
+/// @brief Generates a random user id
+static void GenerateUid(void* /* dummy */)
+{
+    // Create random userId
+    Random random;
+    int uid = random.NextInt(10000);
+    // Format + assign it
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:573]:
```suggestion
```

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 3]
```diff
@@ -535,13 +553,36 @@ static void SelectBlackBox(void* /* dummy */)
...
+    sprintf(m_userId, "%04d", uid);
+    // Display it
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:575]:
```suggestion
```

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 4]
```diff
@@ -535,13 +553,36 @@ static void SelectBlackBox(void* /* dummy */)
...
 
     // Only after validation copy into the actual variable
     strcpy_s(m_blackBoxFilePath, BLACKBOX_PATH_SIZE, buf);
     m_blackBoxChosen = true;
 }
 
+/// @brief Generates a random user id
+static void GenerateUid(void* /* dummy */)
+{
+    // Create random userId
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:570]: Add your comments above a block of code, but not above every single line. In my opinion no comments are needed at all for those lines, everyone will be able to quickly understand what the code does by the code itself.

**Review by larsdekwant — CHANGES_REQUESTED** [review]

**Review by larsdekwant — APPROVED** [review]

**Pull request merged**